### ISSUE TITLE
ODE Status Monitoring

### ISSUE THREAD

**Issue by 11718pauld** [JuliaDiffEq/Sundials.jl#40]
In Matlab we have 'odeprint' which is very handy in monitoring the Progress of ode, is there any equivalent thing for Sundials-Julia?

**Comment by ViralBShah** [comment]
I am not sure if there is anything at the moment. However, in the API, I believe it should not be difficult to provide a way to register a post-processing function that gets called after evaluation at every time step.

**Comment by 11718pauld** [comment]
odeprint is really very handy in monitoring the status of ode solver, when sizeof(y_dot) is large and solver takes whole lot of time to solve.

**Comment by ViralBShah** [comment]
It would be great if you can create a PR. That will be the fastest way to get the functionality in. Otherwise, I am sue one of the regular Sundials contributors will get around to it.

**Comment by acroy** [comment]
I don't know if there is anything related in the Sundials API, but we could certainly call some monitoring function in our high-level functions `cvode` and `idasol`. For example, we could have
```
abstract ProgressMonitor

# called to initialize progress monitor, eg resetting step counter
function init!(p::ProgressMonitor, tspan, y0)
end

# called after each successful step
function step_done(p::ProgressMonitor, t, y0)
end

# default monitor which does nothing
type DoNothing <: ProgressMonitor
end

# simple monitor which prints each step with the result
type PrintProgress <: ProgressMonitor
end

function step_done(p::PrintProgress, t, y0)
  println("@t=$(t): $(y0)")   # this should give some nicely formatted output
end
```
And then we make `cvode` and `idasol` taking an additional argument of type `ProgressMonitor` with the default `DoNothing()`. Users could implement more sophisticated monitors using a progress bar or such.

**Comment by ChrisRackauckas** [comment]
You can do this with callback functions checking current time vs ending time. In fact, using that you can link it into Juno's progressbar pretty easily without adding a dependency (using Main as a nifty trick since Juno always imports its packages to main). I'll be wrapping Sundials in DifferentialEquations.jl pretty soon and I'll make the callback function do this by default.

**Comment by ChrisRackauckas** [comment]
I actually can't find callback functions in the cvode documentation.  Does anyone know how to add a callback function?

**Comment by ChrisRackauckas** [comment]
This will be fixed with Julia v1.0. We'll plug into the base logging system which allows for this information. It separates the information from the display, so from there supporting Juno or console displays would simply be about someone else building a package which supports it (and Juno is already commited to changing their progress displays to this)

**Comment by ChrisRackauckas** [comment]
The common interface progress bar support has been added, so `sol  = solve(prob,CVODE_BDF(),progress=true)` etc. all works.
